plugins/SlicerT/SlicerTUI.h

```diff
@@ -0,0 +1,92 @@
+/*
+ * SlicerTUI.h - declaration of class SlicerTUI
+ *
+ * Copyright (c) 2006-2008 Daniel Kauss Serna <daniel.kauss.serna@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef SLICERT_UI_H
+#define SLICERT_UI_H
+
+#include <QPushButton>
+
+#include "ComboBox.h"
+#include "Instrument.h"
+#include "InstrumentView.h"
+#include "Knob.h"
+#include "LcdSpinBox.h"
+#include "LedCheckBox.h"
+#include "PixmapButton.h"
+#include "WaveForm.h"
+
+namespace lmms {
+
+class SlicerT;
+
+namespace gui {
+
+// style knob, defined in data/themes/default/style.css#L949
+class SlicerTKnob : public Knob
```

Comment:
Any reason for this? I saw that a lot of the plugins handle the knobs like I do, so I did that. I think its easier to read that way.

Reply:
I figured that since your custom slicer knob is basically the same as the original know with a few settings applied, it's easier for you and you don't have to make another whole class just for that.

Reply:
How would this work with the style? I'm pretty sure that the css needs a class name to work, could be wrong though. Either way, I think declaring the knob at the top of the header is better than somewhere hidden in a function.

Reply:
> How would this work with the style? I'm pretty sure that the css needs a class name to work, could be wrong though.

Let me know if I'm missing something, but it doesn't seem like the other classes have a custom ``Knob`` class declared when they want to style them through the CSS stylesheets, so I think it should still be styled regardless if it's in the corresponding class. The CSS changes you made also don't seem to reference ``SlicerTKnob`` in any way. In any case, do try to remove ``SlicerTKnob`` and see if your ``Knob``'s get styled or not, I'm not too sure myself.

Reply:
You are right, I mistakenly thought that I put the knob's class name in the CSS. If I just use a normal styled knob the style gets applied correctly. No problem then, I'll move it to a function.

Reply:
I thought quite a lot if moving the knobs to a function was a good idea, but I ultimately came to the conclusion that having it in a class is better. This mostly comes down too having all the knobs be initialized like all the other members in the member initializer lists. This also means that you can see clearly in the header that the knobs used aren't just the normal ones, but have some changes applied. I also don't really like the syntax of creating a knob through a function.

Reply:
> This mostly comes down too having all the knobs be initialized like all the other members in the member initializer lists.

They don't even need to be initialized there and stored as a member in the class because it does not seem like they are referenced anywhere else in ``SlicerTUI`` but in the constructor. If there only purpose is to be created, doing something like:

```cpp
auto noteThresholdKnob = createKnob(this);
noteThresholdKnob->move(9, 195);
noteThresholdKnob->setToolTip(tr("Threshold used for slicing"));
noteThresholdKnob->setLabel(tr("Threshold"));
noteThresholdKnob->setModel(&m_slicerTParent->m_noteThreshold);
```

Directly in the constructor seems a lot better, where ``this`` is the parent of the ``Knob``. This brings up another problem. When you constructed a ``Knob``, you made a ``QWidget``. Qt has it's own memory management involving parents and children of said parents, so when you created the ``Knob`` on the stack and gave it a parent, we now have problems, like who should destroy the ``Knob``, Qt's own memory management system or ``~SlicerTUI``?

The resolution is to use ``new`` instead and not store the ``Knob`` as a member on the stack (at least store it as a pointer instead, but as mentioned, it doesn't even seem like you have to).

> This also means that you can see clearly in the header that the knobs used aren't just the normal ones, but have some changes applied.

The header shouldn't even need to have references to them. In the case that it would, I think calling a function called ``createKnob`` or ``createSlicerKnob`` (that is inside the ``SlicerTUI`` class by the way, making it even clearer) is clear enough.

>  I also don't really like the syntax of creating a knob through a function.

Huh? Why is that? I suggested a function mainly because you shouldn't duplicate the same changes you make to each of them.

Reply:
> Qt has it's own memory management involving parents and children of said parents, so when you created the Knob on the stack and gave it a parent, we now have problems, like who should destroy the Knob, Qt's own memory management system or ~SlicerTUI?

Sorry I didn't know this, I'm not very familiar with Qt. You make some good points, so I have added the function. This is done.